lib/models/addon.js

```diff
@@ -248,8 +261,39 @@ var Addon = CoreObject.extend({
     this.initializeAddons();
 
     var invokeArguments = args || [];
-
+    var parent = this;
     return this.addons.map(function(addon) {
+      if (methodName === 'treeFor') {
```

Comment:
I'm tempted that, if the addon has non-default implementations for these hooks, we just build it. (Rather than looking it up.) I'd guess that this still gets us the majority of the win and _also_ makes it less likely to break things.